lightning/src/util/persist.rs

```diff
@@ -243,10 +250,886 @@ where
+	}
+}
+
+impl TryFrom<&MonitorName> for OutPoint {
+	type Error = io::Error;
+
+	fn try_from(value: &MonitorName) -> Result<Self, io::Error> {
+		MonitorName::do_try_into_outpoint(&value.0)
+	}
+}
+
+impl From<OutPoint> for MonitorName {
+	fn from(value: OutPoint) -> Self {
+		MonitorName(format!("{}_{}", value.txid.to_hex(), value.index))
+	}
+}
+
+/// A struct representing a name for an update.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct UpdateName(u64, String);
+
+impl UpdateName {
+	/// Constructs an [`UpdateName`], after verifying that an update sequence ID
+	/// can be derived from the given `name`.
+	pub fn new(name: String) -> Result<Self, io::Error> {
+		match name.parse::<u64>() {
+			Ok(u) => Ok(u.into()),
+			Err(_) => Err(io::Error::new(
+				io::ErrorKind::InvalidData,
+				"cannot parse u64 from update name",
+			)),
+		}
+	}
+
+	/// Convert this monitor update name to a &str
+	pub fn as_str(&self) -> &str {
+		&self.1
+	}
+
+	/// Left-pad the sequential update id for string representation
+	fn left_padded_string(value: u64) -> String {
+		format!("{:0>20}", value)
+	}
+}
+
+impl From<u64> for UpdateName {
+	fn from(value: u64) -> Self {
+		Self(value, UpdateName::left_padded_string(value))
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::chain::chainmonitor::Persist;
+	use crate::chain::ChannelMonitorUpdateStatus;
+	use crate::events::{ClosureReason, MessageSendEventsProvider};
+	use crate::ln::functional_test_utils::*;
+	use crate::util::test_utils::{self, TestLogger, TestStore};
+	use crate::{check_added_monitors, check_closed_broadcast};
+
+	const EXPECTED_UPDATES_PER_PAYMENT: u64 = 5;
+
+	fn default_test_persister<ES: Deref + Clone, SP: Deref + Clone>(
+		read_only: bool,
+		entropy_source: ES,
+		signer_provider: SP,
+	) -> MonitorUpdatingPersister<Box<TestStore>, Box<TestLogger>, ES, SP>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+	{
+		MonitorUpdatingPersister {
+			kv_store: Box::new(TestStore::new(read_only)),
+			logger: Box::new(TestLogger::new()),
+			maximum_pending_updates: 11,
+			entropy_source,
+			signer_provider,
+		}
+	}
+
+	#[test]
+	fn test_update_from_u64_and_as_str_works() {
+		let cases = [
+			(0, "00000000000000000000"),
+			(21, "00000000000000000021"),
+			(u64::MAX, "18446744073709551615"),
+		];
+		for (input, expected) in &cases {
+			let update_name = UpdateName::from(*input);
+			assert_eq!(update_name.as_str(), *expected);
+		}
+	}
+
+	#[test]
+	fn test_update_from_bad_string_fails() {
+		let cases = [
+			("deadbeef".to_string(), "cannot parse u64"),
+			("-0000000000000000001".to_string(), "cannot parse u64"),
+		];
+		for (input, expected) in &cases {
+			let got = UpdateName::new(input.to_owned())
+				.expect_err("bad update name accepted")
+				.to_string();
+			assert!(
+				got.contains(*expected),
+				"wrong update name error\nexpected: {}\ngot: {}",
+				expected,
+				got
+			);
```

Comment:
Could we use `assert_eq!` here with the full error string instead of crafting our own failure message? I think the test name gives enough context about what is being tested.

Reply:
I did it this way because I'm using `contains`, not equality. So what happens when you don't find the test string is that it panics, so I added the extra context. The more obvious thing to do would be to test equality of the whole error string (this used to be an error enum but I was encouraged to ditch it). But that's gonna be even more brittle unless I make the error message a constant, which in turn seems like overkill.

I think what probably makes the most sense is to just confirm `is_err()` and leave it at that. Doesn't really matter whether we're erroring for exactly the right reason / saying it the right way, right?